torch/jit/__init__.py

```diff
@@ -699,9 +701,29 @@ def script_method(fn):
     #
     # createResolutionCallback internally adds 1 to get us to the scope of this
     # function (the calling function). Adding 2 gets us to the proper surrounding scope.
-    rcb = createResolutionCallback(frames_up=2)
+    if _rcb is None:
+        _rcb = createResolutionCallback(frames_up=2)
     ast = get_jit_ast(fn, is_method=True)
-    return ScriptMethodStub(rcb, ast, fn)
+    return ScriptMethodStub(_rcb, ast, fn)
+
+
+WeakMethodStub = namedtuple('WeakMethodStub', ('resolution_callback', 'original_method'))
+
+
+def weak_script_method(fn):
+    if _enabled:
+        weak_script_methods[fn] = WeakMethodStub(createResolutionCallback(frames_up=2), fn)
+    return fn
+
+
+def _try_get_weak_module(mod):
+    weak_mods_of_class = weak_modules.get(mod.__class__)
+    if weak_mods_of_class is not None:
+        weak_mod = weak_mods_of_class.get(mod)
```

Comment:
What is going on here? I can't follow the logic and have no clue what the class is for any of these types.